Implement Bloom Filter Index as an another indexing strategy for tuples stored in StorageBlock.

saketj opened the pull request:
This PR adds support for a fully functional BloomFilter index. The following SQL statement can be used to create a bloom filter index on a given table:

```
CREATE TABLE foo(col1 INTEGER NOT NULL, col2 INTEGER NOT NULL);
CREATE INDEX bloom_index ON foo(col1) USING BLOOMFILTER (SIZE 10000);
```
In the above statement, `SIZE` is an optional parameter for this index, with `num_hashes` and `projected_element_count` being other properties that can also be specified for a bloom filter index. 

This PR also modifies SubBlockTypeRegistry and adds a function EstimateBytesPerBlock() that can now be used by index sub-blocks like BloomFilter, SMA, etc. to specify a constant space size that they want to occupy within StorageBlock. Earlier, EstimateBytesPerTuples() was the only alternative to specify the space occupied by index sub-blocks within the StorageBlock as a function of number of tuples. This led to space inefficiencies for index sub-blocks that only needed fixed/constant sizes. The EstimateBytesPerBlock() now solves this problem.

The unit tests for this PR can be found in following files:
- storage/tests/BloomFilterIndexSubBlock_unittest.cpp
- query_optimizer/tests/execution_generator/Index.test
- query_optimizer/tests/logical_generator/Index.test
- query_optimizer/tests/physical_generator/Index.test
- query_optimizer/tests/resolver/Index.test

zuyu commented:
@cramja The CI is ok. We just have a known, but frequent-than-expected block counting failure in a newly-added unit test.

cramja commented:
@zuyu Yes, I noticed that it happened twice in some of the recent builds that Rogers was working on related to \dt.

cramja commented on storage/BloomFilterIndexSubBlock.hpp:
```diff
@@ -0,0 +1,184 @@
+    kSelectivityNone
+  };
+
+  static const std::uint64_t BLOOM_FILTER_SEED = 0xA5A5A5A55A5A5A5AULL;
```
Constants are usually written in the form `kConstantName`.

saketj replied:
Refactored.

cramja commented on storage/BloomFilterIndexSubBlock.cpp:
```diff
@@ -0,0 +1,271 @@
+
+std::size_t BloomFilterIndexSubBlock::EstimateBytesPerTuple(const CatalogRelationSchema &relation,
+                                                            const IndexSubBlockDescription &description) {
+  DEBUG_ASSERT(DescriptionIsValid(relation, description));
```
Please use google's CHECK macros instead of DEBUG_ASSERT

saketj replied:
Refactored.

cramja commented on storage/BloomFilterIndexSubBlock.cpp:
```diff
@@ -0,0 +1,271 @@
+std::size_t BloomFilterIndexSubBlock::EstimateBytesPerBlock(const CatalogRelationSchema &relation,
+                                                            const IndexSubBlockDescription &description) {
+  // Note: This function is only invoked when EstimateBytesPerTuple() returns zero.
+  DEBUG_ASSERT(DescriptionIsValid(relation, description));
```
Please use google's CHECK macros instead of DEBUG_ASSERT

saketj replied:
Refactored to use CHECK instead of DEBUG_ASSERT.

cramja commented on storage/BloomFilterIndexSubBlock.cpp:
```diff
@@ -0,0 +1,271 @@
+    // Best-case scenario, when a BloomFilter index allows to skip a block entirely.
+    // A new tuple ID sequence is initialized to false for all values.
+    return new TupleIdSequence(tuple_store_.numTuples());
+  } else if (selectivity == BloomFilterSelectivity::kSelectivityAll) {
```
A bloomfilter can discover that all tuples match?

cramja replied:
Oh I see, the definite for `all` isn't exactly what I thought it meant. It seems to mean that:

```c++
 // Bloom filter hit. Can be a false positive, a scan is required here.
 return BloomFilterSelectivity::kSelectivityAll;
```

cramja commented on storage/BloomFilterIndexSubBlock.cpp:
```diff
@@ -0,0 +1,271 @@
+  }
+  // Bloom filter index can be currently only used to evaluate
+  // equality matching comparison predicates with static values.
+  if (predicate.getComparison().getComparisonID() == ComparisonID::kEqual
```
Does the bloom filter handle the case where the static value is on the left hand side?

saketj replied:
No, currently bloom filter only support static values on the right. Thanks, have updated the comment to indicate that.

cramja commented on storage/BloomFilterIndexSubBlock.hpp:
```diff
@@ -0,0 +1,184 @@
+**/
+class BloomFilterIndexSubBlock : public IndexSubBlock {
+ public:
+  enum class BloomFilterSelectivity {
```
Please add comments here describing what the enums mean. It would especially help with the All case since that was unclear until reading into the code.

saketj replied:
Added comments now.

cramja commented on storage/BloomFilterIndexSubBlock.hpp:
```diff
@@ -0,0 +1,184 @@
+    return kBloomFilter;
+  }
+
+  bool supportsAdHocAdd() const override {
```
This does support, no?

saketj replied:
Thanks, yeah adhoc inserts are supported and adhoc removes are not. Updated it.

cramja commented on storage/BloomFilterIndexSubBlock.hpp:
```diff
@@ -0,0 +1,184 @@
+   * @brief Get the selectivity provided by the index for a given predicate.
+   *        There are only two selectivity levels returned by this function-
+   *        either a zero(none) selectivity for bloom filter miss
+   *        or a one(all) selectivity for bloom filter hit.
```
Might want to add that a 'hit' means that the predicate possibly returns tuples.

saketj replied:
Updated.

cramja commented on storage/CSBTreeIndexSubBlock.cpp:
```diff
@@ -353,6 +355,12 @@ std::size_t CSBTreeIndexSubBlock::EstimateBytesPerTuple(
+std::size_t CSBTreeIndexSubBlock::EstimateBytesPerBlock(
+    const CatalogRelationSchema &relation,
+    const IndexSubBlockDescription &description) {
+  return 0;
```
Maybe we should make a constant in `IndexSubBlock.hpp` to describe when an `EstimateSize` function is not applicable...

`size_t kSizeNotApplicable = 0` ?

saketj replied:
@cramja Sure, it makes sense to add a boolean flag indicating this. But I am still struggling to find a proper place to put this variable. IndexSubBlock.hpp seems to only define virtual functions. and EstimateBytesPerTuple() and EstimateBytesPerBlock() are not even defined in the base class itself. Maybe we can put it in each derived class file.

cramja commented on storage/SMAIndexSubBlock.cpp:
```diff
@@ -488,6 +490,12 @@ std::size_t SMAIndexSubBlock::EstimateBytesPerTuple(
+std::size_t SMAIndexSubBlock::EstimateBytesPerBlock(
+    const CatalogRelationSchema &relation,
+    const IndexSubBlockDescription &description) {
+  return 0;
```
Guess I gotta go make the SMA better now.. That's been on my TODO for a while, ty for adding this.

cramja commented on storage/StorageBlockLayout.cpp:
```diff
@@ -114,7 +116,15 @@ void StorageBlockLayout::finalize() {
+      // Estimated size to be consumed was specified per tuple.
+      index_size = (sub_block_space * index_size_factors[index_num]) / estimated_bytes_per_tuple_;
+    } else {
+      // Some indicies define total size per block, instead of defining size per tuple.
```
indicies -> indices

saketj replied:
Fixed.

cramja commented on storage/SubBlockTypeRegistry.cpp:
```diff
@@ -106,4 +108,17 @@ std::size_t SubBlockTypeRegistry::EstimateBytesPerTupleForIndex(
 
+std::size_t SubBlockTypeRegistry::EstimateBytesPerBlockForIndex(const CatalogRelationSchema &relation,
+                                                                const IndexSubBlockDescription &description) {
+  DEBUG_ASSERT(description.IsInitialized());
```
Please use google's CHECK macros instead of DEBUG_ASSERT

saketj replied:
Refactored to use CHECK instead of DEBUG_ASSERT.

cramja replied:
Thankyou for using CHECK instead of DEBUG_ASSERT

cramja commented on storage/SubBlockTypeRegistry.cpp:
```diff
@@ -106,4 +108,17 @@ std::size_t SubBlockTypeRegistry::EstimateBytesPerTupleForIndex(
+  IndexEstimateBytesPerBlockFunction>::const_iterator
+  it = Instance()->index_estimate_bytes_per_block_functions_.find(
+      static_cast<IndexTypeIntegral>(description.sub_block_type()));
+  DEBUG_ASSERT(it != Instance()->index_estimate_bytes_per_block_functions_.end());
```
Please use google's CHECK macros instead of DEBUG_ASSERT

saketj replied:
Refactored to use CHECK instead of DEBUG_ASSERT.

cramja commented on utility/BloomFilter.hpp:
```diff
@@ -0,0 +1,232 @@
+ **/
+class BloomFilter {
+ public:
+  static const uint32_t NUM_BITS_PER_BYTE = 8;
```
Constants are usually of the form `kConstantName`.

saketj replied:
Fixed.

cramja commented:
@saketj In general, looks very clear/clean. I like that we have integration tests as well as unit tests for the BloomFilterSubBlock. 

I did not look over code for the  utility/BloomFilter implementation. Maybe we could walk through that in person?

saketj commented:
@cramja  Thanks Marc for the comments. I have incorporated the changes except for the EstimateSize flag. Maybe we can discuss more on that.

cramja commented on utility/CMakeLists.txt:
```diff
@@ -263,7 +267,11 @@ target_link_libraries(quickstep_utility
                       quickstep_utility_Alignment
                       quickstep_utility_BitManipulation
                       quickstep_utility_BitVector
+<<<<<<< HEAD
```
I think these are left over from a merge?

saketj replied:
Removed.

cramja commented on utility/CMakeLists.txt:
```diff
@@ -199,8 +200,11 @@ target_link_libraries(quickstep_utility_BitVector
+                      quickstep_utility_Macros)
 target_link_libraries(quickstep_utility_CalculateInstalledMemory
-                      glog)
+                      glog)                      
```
Extra whitespace introduced.

saketj replied:
Removed.

cramja commented:
@saketj Good job with this PR. At the current time, the only thing that worries me is that there is no test for the BloomFilter class int `utility`. It could be a simple test, just to test the `contains` and `insert` methods.

saketj commented:
@cramja Thanks Marc for the comments. I will also add the tests for the BloomFilter class to this PR.

saketj commented:
@cramja Thanks for the comments earlier. I have added the unit tests for BloomFilter utility class now and also refactored to return a constant from EstimateBytesPerTuple() from fixed-size index sub-blocks.

cramja commented:
Great, when travis returns we can go ahead and merge.

cramja commented:
@saketj Looks good, merging.
